File: digio-account/src/main/java/in/digio/account/filter/TenantFilter.java
package in.digio.account.filter;

import in.digio.account.dto.WorkSpaceRequest;
import in.digio.account.dto.WorkSpaceResponse;
import in.digio.account.service.WorkSpaceService;
import in.digio.auth.model.user.User;
import in.digio.core.multitenancy.TenantContext;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.core.annotation.Order;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

import static in.digio.auth.config.SecurityConfiguration.*;

@Order(2)
@Component
@RequiredArgsConstructor
public class TenantFilter extends OncePerRequestFilter {

    private final WorkSpaceService workSpaceService;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        if (Arrays.stream(OPEN_APIS).anyMatch(api -> request.getServletPath().contains(api)) || request.getServletPath().contains(USER_APIS)) {
            filterChain.doFilter(request, response);
            return;
        }
        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        TenantContext.setCurrentTenant(user.getId());

        if (request.getServletPath().contains(ACCOUNT_APIS)) {
            filterChain.doFilter(request, response);
        } else {
            // Set current workspace id.
            List<WorkSpaceResponse> workSpaces = workSpaceService.getWorkSpaces();
            if (workSpaces.isEmpty()) {
                String fullName = user.getFullName();
                WorkSpaceResponse workSpace = workSpaceService.createWorkSpace(WorkSpaceRequest.builder().name(fullName.isEmpty() ? "Default WorkSpace" : fullName).build());
                TenantContext.setCurrentTenant(workSpace.getId());
            } else {
                TenantContext.setCurrentTenant(workSpaces.get(0).getId());
            }
            filterChain.doFilter(request, response);
        }
        TenantContext.unload();
    }
}

File: core/src/main/kotlin/com/ampairs/core/multitenancy/TenantFilter.kt
package com.ampairs.core.multitenancy

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.core.annotation.Order
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
@Order(1)
@ConditionalOnProperty(name = ["application.multitenancy.enabled"], havingValue = "true", matchIfMissing = true)
class TenantFilter : OncePerRequestFilter() {

    private val logger = LoggerFactory.getLogger(TenantFilter::class.java)

    companion object {
        const val TENANT_HEADER = "X-Tenant-ID"
        const val TENANT_PARAM = "tenantId"
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        try {
            val tenantId = extractTenantId(request)

            if (tenantId != null) {
                TenantContext.setCurrentTenant(tenantId)
                logger.debug("Tenant context set for request: {} to {}", request.requestURI, tenantId)
            } else {
                logger.debug("No tenant ID found for request: {}", request.requestURI)
            }

            filterChain.doFilter(request, response)
        } finally {
            TenantContext.clearTenantContext()
            logger.debug("Tenant context cleared after request: {}", request.requestURI)
        }
    }

    private fun extractTenantId(request: HttpServletRequest): String? {
        // Try header first
        request.getHeader(TENANT_HEADER)?.let { return it }

        // Try request parameter
        request.getParameter(TENANT_PARAM)?.let { return it }

        // Try to extract from authentication context
        SecurityContextHolder.getContext().authentication?.let { auth ->
            when {
                auth.principal is TenantAware -> {
                    return (auth.principal as TenantAware).getTenantId()
                }

                auth.details is Map<*, *> -> {
                    val details = auth.details as Map<String, Any>
                    return details["tenantId"] as? String
                }
            }
        }

        // Try to extract from subdomain (if applicable)
        extractTenantFromSubdomain(request)?.let { return it }

        return null
    }

    private fun extractTenantFromSubdomain(request: HttpServletRequest): String? {
        val serverName = request.serverName
        if (serverName.contains(".")) {
            val parts = serverName.split(".")
            if (parts.size >= 3) { // subdomain.domain.tld
                val subdomain = parts[0]
                // Only consider valid tenant subdomains (not www, api, etc.)
                if (subdomain.matches(Regex("[a-zA-Z0-9-]{3,}"))) {
                    return subdomain
                }
            }
        }
        return null
    }
}

interface TenantAware {
    fun getTenantId(): String?
}